### PR TITLE
Gilded Rose Solution

### PR DESCRIPTION
**Pull request by fgs-pwright**

### CODE REVIEW

File: src/GildedRose.Console/Program.cs [hunk 1]
```diff
@@ -2,123 +2,39 @@
 
 namespace GildedRose.Console
 {
-    class Program
+    public class Program
     {
-        IList<Item> Items;
-        static void Main(string[] args)
+        private IList<Item> _items;
+        public static void Main(string[] args)
```
Comment (FGS-BMarkham) [inline src/GildedRose.Console/Program.cs]: As is usually the case for `Main` methods with too much logic, this looks really difficult to test. Consider finding ways to extract more of this out. Ideally, a `Main` method should only have the barest amount of logic needed to hand off control to (testable) components.

File: src/GildedRose.Tests/AgedBrieTests.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        {
+            Name = "Aged Brie";
+        }
+
+        [Test]
+        public void UpdateAgedBrieQuality_GivenPositiveSellIn_IncreasesQualityByOne()
+        {
+            var sellIn = Fixture.Create<int>();
+            var quality = Fixture.CreateInRange<int>(Inventory.MINQUALITY, Inventory.MAXQUALITY - 1);
+            var brie = new Item { Name = Name, SellIn = sellIn, Quality = quality };
```
Comment (FGS-BMarkham) [inline src/GildedRose.Tests/AgedBrieTests.cs]: This looks really similar to setup logic in the next test. Can you DRY them up?

File: src/GildedRose.Tests/AutoFixtureExtension.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
+﻿using Ploeh.AutoFixture;
+
+namespace GildedRose.Tests
+{
+    public static class AutoFixtureExtension
+    {
+        public static T CreateInRange<T>(this Fixture fixture, int lowerLimit, int upperLimit)
+        {
+            fixture.Customizations.Add(new RandomNumericSequenceGenerator(lowerLimit, upperLimit));
```
Comment (FGS-BMarkham) [inline src/GildedRose.Tests/AutoFixtureExtension.cs]: Good use of AutoFixture's API to customize it. Don't be afraid to look for these kinds of extensibility points on other libraries we use, too.

File: src/GildedRose.Tests/BackstagePassTests.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+        [TestCase(BackstagePassUpdater.MANYDAYS + 1, byte.MaxValue, 1)]
+        [TestCase(BackstagePassUpdater.FEWDAYS + 1, BackstagePassUpdater.MANYDAYS, 2)]
+        [TestCase(BackstagePassUpdater.NODAYS + 1, BackstagePassUpdater.FEWDAYS, 3)]
+        public void UpdateBackstagePassQuality_GivenSellInWithinRange_IncreasesQualityBy(
+            int minDays, int maxDays, int expectedChange)
+        {
+            var sellIn = Fixture.CreateInRange<int>(minDays, maxDays);
+            var quality = Fixture.CreateInRange<int>(Inventory.MINQUALITY, Inventory.MAXQUALITY - expectedChange);
+            var passes = new Item { Name = Name, SellIn = sellIn, Quality = quality };
```
Comment (FGS-BMarkham) [inline src/GildedRose.Tests/BackstagePassTests.cs]: You can improve readability by extracting these three lines into a function with a name representing the state that they setup. For example,
```
var passes = Given_SellInWithinRange(minDays, maxDays, expectedChange);
```
Look for these opportunities in your other tests as well.

Also, I'm not sure the variable name "passes" makes sense. Please consider renaming it here and elsewhere.

File: src/GildedRose.Console/Inventory.cs [hunk 5]
```diff
@@ -0,0 +1,33 @@
+﻿using System.Collections.Generic;
+
+namespace GildedRose.Console
+{
+    public class Inventory
+    {
+        public const int MINQUALITY = 0;
+        public const int MAXQUALITY = 50;
```
Comment (FGS-CThomas) [inline src/GildedRose.Console/Inventory.cs]: WHICH IS A PERFECTLY ACCEPTABLE LIFE CHOICE.

Comment (FGS-JTKuhn) [inline src/GildedRose.Console/Inventory.cs]: >WHICH_IS_A_PERFECTLY_ACCEPTABLE_LIFE_CHOICE
ftfy

Comment (fgs-dbudwin) [inline src/GildedRose.Console/Inventory.cs]: @FGS-BMarkham be all like...

File: src/GildedRose.Console/Item.cs [hunk 6]
```diff
@@ -0,0 +1,11 @@
+﻿namespace GildedRose.Console
+{
+    public class Item
+    {
+        public string Name { get; set; }
+
+        public int SellIn { get; set; }
```
Comment (fgs-pwright) [inline src/GildedRose.Console/Item.cs]: I agree, but I left it as this on account of the goblin.

> ...do not alter the Item class or Items property as those belong to the goblin in the corner who will insta-rage and one-shot you as he doesn't believe in shared code ownership...

File: src/GildedRose.Console/Program.cs [hunk 7]
```diff
@@ -2,123 +2,39 @@
...
-                                                      Quality = 20
-                                                  },
-                                              new Item {Name = "Conjured Mana Cake", SellIn = 3, Quality = 6}
-                                          }
-
-                          };
+            {
+                _items = new List<Item>
+                {
+                    new Item { Name = "+5 Dexterity Vest", SellIn = 10, Quality = 20 },
```
Comment (fgs-pwright) [inline src/GildedRose.Console/Program.cs]: When you talk about injecting the factory into the constructor, are you talking about doing something like this?

```cs
public class Inventory
{
    // ...
    private ItemFactory _itemFactory;

    public Inventory()
        : this(new ItemFactory())
    {
    }

    private Inventory(ItemFactory itemFactory)
    {
        _itemFactory = itemFactory;
        // ...
    }
    
    // ...
}
```

File: src/GildedRose.Console/Program.cs (hunk 1 above)
Comment (fgs-pwright) [inline src/GildedRose.Console/Program.cs]: This is a part of the assignment I'm not really sure I completely understand, and I think it's because before the refactor, the console application didn't really do anything that would be useful to a user.  I'm not sure if I should be building a command-line interface that would allow a user to call methods I've written in other classes, or if I should just be extracting the procedural code from the `Main` method into some other `Run` method that performs the same operations that were in `Main` before.

Comment (fgs-dbudwin) [inline src/GildedRose.Console/Program.cs]: > I should just be extracting the procedural code from the Main method into some other Run method that performs the same operations that were in Main before.

☝️ This is what you should do, although I wouldn't call it `Run`.  Bruce's point is that the `Main` method needs to me as minimal as possible and only do enough to get the program running.

File: src/GildedRose.Console/Program.cs (hunk 7 above)
Comment (fgs-dbudwin) [inline src/GildedRose.Console/Program.cs]: Kind of...since we don't have you using a dependency injection framework yet, we'll have to use a form of dependency injection--constructor injection.  I would rephrase your above example to `public Inventory(ItemFactory itemFactory)`.  This will still result in a `new` call somewhere, but at least we get to push it up the chain a bit more.

File: src/GildedRose.Console/Program.cs (hunk 1 above)
Comment (fgs-pwright) [inline src/GildedRose.Console/Program.cs]: Ok, got it.  I'll put something together for this and push again.

Comment (fgs-pwright) [inline src/GildedRose.Console/Program.cs]: > the Main method needs to me as minimal as possible and only do enough to get the program running.
With this in mind, I created an `InventoryManager` class and extracted the logic that was previously in `Main` to it.  This class could be extended to do other possibly more useful things that are outside of the scope of the kata.

Comment (fgs-dbudwin) [inline src/GildedRose.Console/Program.cs]: Cool!  However, I would avoid naming class with the term "Manager."  See the following:

https://blog.codinghorror.com/i-shall-call-it-somethingmanager/

Comment (fgs-pwright) [inline src/GildedRose.Console/Program.cs]: Ah, thanks.  Got it.

File: src/GildedRose.Console/InventoryQualityController.cs [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+        public void Start()
+        {
+            Initialize();
+            _inventory.AgeAlterableItems();
+        }
+
+        private void Initialize()
+        {
+            _inventory.AddItem("+5 Dexterity Vest", 10, 20);
```
Comment (fgs-dbudwin) [inline src/GildedRose.Console/InventoryQualityController.cs]: Since this strings are repeated a few times in the program and subject to typos, it may be best to move them to a common location where they can be referred to.

File: src/GildedRose.Console/Program.cs [hunk 9]
```diff
@@ -1,124 +1,18 @@
...
-                                {
-                                    Items[i].Quality = Items[i].Quality + 1;
-                                }
-                            }
-                        }
-                    }
-                }
+            var inventory = new Inventory(new ItemFactory(), new UpdaterBuilder());
+            var inventoryQualityController = new InventoryQualityController(inventory);
+            inventoryQualityController.Start();
```
Comment (FGS-BMarkham) [inline src/GildedRose.Console/Program.cs]: This is way better, good job!

File: src/GildedRose.Console/InventoryQualityController.cs [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+        private Inventory _inventory;
+
+        public InventoryQualityController(Inventory inventory)
+        {
+            _inventory = inventory;
+        }
+
+        public void Start()
+        {
+            Initialize();
```
Comment (fgs-dbudwin) [inline src/GildedRose.Console/InventoryQualityController.cs]: I would call this something else like `AddItemsToInventory`

File: src/GildedRose.Console/Updater.cs [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+        public Updater(Item item)
+        {
+            Item = item;
+        }
+
+        public void AgeItem()
+        {
+            var changeAmount = CalculateQualityChange();
+            ChangeQualityBy(changeAmount);
```
Comment (fgs-dbudwin) [inline src/GildedRose.Console/Updater.cs]: This seems like a SRP violation.  What does aging an item have to do with changing the quality?  If I called this function, based on its name, I wouldn't expect the quality to also change.

File: src/GildedRose.Tests/BackstagePassTests.cs [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        [SetUp]
+        public void SetName()
+        {
+            Item.Name = Fixture.Create("Backstage passes");
+        }
+
+        [TestCase(BackstagePassUpdater.MANY_DAYS + 1, byte.MaxValue, 1)]
+        [TestCase(BackstagePassUpdater.FEW_DAYS + 1, BackstagePassUpdater.MANY_DAYS, 2)]
+        [TestCase(BackstagePassUpdater.NO_DAYS + 1, BackstagePassUpdater.FEW_DAYS, 3)]
```
Comment (FGS-BMarkham) [inline src/GildedRose.Tests/BackstagePassTests.cs]: It is not obvious to me why `1` is being added to the `minDays` parameter. Is this a commonality that can be pushed into the test?

Please also note that test runners will not be able to list these test cases with the identifiers of constants in their names, which will actually make them very difficult to discern. One way to address this would be to switch to an `enum` based configuration. Another way would be to use the parameter of `TestCase` that allows you to name the test case with prose. Please considering addressing this for this test any others that might be like it.

File: src/GildedRose.Console/Updater.cs (hunk 11 above)
Comment (fgs-dbudwin) [inline src/GildedRose.Console/Updater.cs]: I want to further clarify something here that may be confusing.  Aging an item to me is a reflection on the sell by date, not the item's quality.

Comment (fgs-dbudwin) [inline src/GildedRose.Console/Updater.cs]: A solution could also be to change the name of this function.

File: src/GildedRose.Tests/BackstagePassTests.cs (hunk 12 above)
Comment (fgs-pwright) [inline src/GildedRose.Tests/BackstagePassTests.cs]: I went ahead and named these test cases using language relevant to the item type to make this clearer.

File: src/GildedRose.Console/Inventory.cs [hunk 13]
```diff
@@ -0,0 +1,65 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+
+namespace GildedRose.Console
+{
+    public class Inventory
```
Comment (FGS-BMarkham) [inline src/GildedRose.Console/Inventory.cs]: I am not overly familiar with the constraints of this kata, but I'm not sure that an "Inventory" class needs to exist. In other words: when you remove the things from this class that look like SRP violations to me (the default item population, the constants) - it becomes an anemic wrapper around a list (without even encapsulating said list). That is to say - if those parts are properly extracted, you could just use a list everywhere you are using an `Inventory` right now. What do you think?

File: src/GildedRose.Tests/InventoryTests.cs [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+        [Test]
+        public void NewInventory_HasItemListMatchingDefaultItems()
+        {
+            var defaultItems = new List<Item>();
+            PopulateWithDefaultItems(defaultItems);
+
+            AssertCollectionsAreEquivalent(Inventory.Items, defaultItems);
+        }
+
+        private void PopulateWithDefaultItems(List<Item> items)
```
Comment (FGS-BMarkham) [inline src/GildedRose.Tests/InventoryTests.cs]: Always avoid concrete collection types for parameters, return values, and members.

When selecting the appropriate abstract collection type to use, use an order of precedence like this:
`IEnumerable<T>` > `IReadOnlyCollection<T>` > `ICollection<T>` > `IReadOnlyList<T>` > `IList<T>`

- `IEnumerable<T>` works just fine if you are only using `foreach` or `LINQ` to read through the contents (or project new data).
- Use `IReadOnlyCollection<T>` if the underlying concrete collection type has a `Count` property and you need to work with said value (and relying on LINQ's `.Count()` extension _method_, which is implemented with a `foreach`, is too costly).
- Use `IReadOnlyList<T>` if you wish to retrieve items by their index. Effective LINQ usage makes this rare, so regard this as a code smell.
- Use `ICollection<T>` if you need to add/remove items. Effective LINQ usage makes this rare, so regard this as a code smell.
- Use `IList<T>` if you need to replace/remove items by index. Effective LINQ usage makes this rare, so regard this as a code smell.

(Note that `List<T>` implements _all of these interfaces_.)

Spend some time playing around with these to get more familiar with them. Also, if you haven't seen it yet, Visual Studio's "Object Browser" (`View` -> `Object Browser`) will show you all the types relevant to your current project. It shows type members, along with inherited types.
Consider using it to gander at `mscorlib`'s `System.Collections.*` namespace to get a feel for how these are all related to each other:
image

File: src/GildedRose.Console/Inventory.cs (hunk 13 above)
Comment (fgs-pwright) [inline src/GildedRose.Console/Inventory.cs]: That's a good point, and I think I could make the switch from `Inventory` to an `Item` list within the constraints of the kata.

Figuring out where to put the default item population is what I've struggled with the most.  What would you think of me putting it in my `ItemFactory` class?

```cs
public class ItemFactory
{
    public IEnumerable<Item> CreateDefaultItems() // This is new
    // These are already here:
    public AgedBrieItem CreateAgedBrieItem(string name, int daysToSell, int quality)
    public BackstagePassItem CreateBackstagePassItem(string name, int daysToSell, int quality)
    public ConjuredItem CreateConjuredItem(string name, int daysToSell, int quality)
    public LegendaryItem CreateLegendaryItem(string name, int daysToSell, int quality)
    public StandardItem CreateStandardItem(string name, int daysToSell, int quality)
}
```
Or I could create a special class just for creating the default items, if this looks like it would violate SRP.

Comment (FGS-BMarkham) [inline src/GildedRose.Console/Inventory.cs]: > I could create a special class just for creating the default items, if this looks like it would violate SRP.

Your instinct is correct, it doesn't really fit in the existing factory. I'd put it into a different class. You could call it something like `InventoryFactory`, maybe. ¯\_(ツ)_/¯

File: src/GildedRose.Tests/InventoryTests.cs (hunk 14 above)
Comment (fgs-pwright) [inline src/GildedRose.Tests/InventoryTests.cs]: Thanks for the detail on this.  It's very helpful.
One of the constraints on the kata was that the `items` property of the `Program` class should not be changed, so I left that as an `IList<Item>` but changed collection types to abstract ones in the rest of the code.

File: src/GildedRose.Tests/InventoryFactoryTests.cs [hunk 15]
```diff
@@ -0,0 +1,47 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using GildedRose.Console;
+using NUnit.Framework;
+
+namespace GildedRose.Tests
+{
+    [TestFixture]
+    public class InventoryFactoryTests : InventoryTests
```
Comment (FGS-BMarkham) [inline src/GildedRose.Tests/InventoryFactoryTests.cs]: If you follow my decoupling advice, you'll want to make sure you are using a _mock_ of `IItemFactory` inside the `InventoryFactory` that is being tested here. This will help keep these tests from being too sensitive to how `ItemFactory` (the concrete class) is implemented. The goal being: unit tests, usually, should only be testing the behavior of the subject class - and not second-level (or greater) dependencies.

File: src/GildedRose.Console/ItemFactory.cs [hunk 16]
```diff
@@ -1,28 +1,28 @@
 ﻿namespace GildedRose.Console
 {
-    public class ItemFactory
+    public static class ItemFactory
```
Comment (FGS-BMarkham) [inline src/GildedRose.Console/ItemFactory.cs]: Also, in the spirit of using the least-specific type needed, all of these factory methods can probably have your root `Item` type as their return types.

File: src/GildedRose.Console/InventoryFactory.cs [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+        public static IEnumerable<Item> CreateDefaultItems()
+        {
+            var items = new Collection<Item>();
+            items.Add(ItemFactory.CreateStandardItem("+5 Dexterity Vest", 10, 20));
+            items.Add(ItemFactory.CreateAgedBrieItem("Aged Brie", 2, 0));
+            items.Add(ItemFactory.CreateStandardItem("Elixir of the Mongoose", 5, 7));
+            items.Add(ItemFactory.CreateLegendaryItem("Sulfuras, Hand of Ragnaros", 0, 80));
+            items.Add(ItemFactory.CreateBackstagePassItem("Backstage passes to a TAFKAL80ETC concert", 15, 20));
+            items.Add(ItemFactory.CreateConjuredItem("Conjured Mana Cake", 3, 6));
+            return items;
```
Comment (fgs-pwright) [inline src/GildedRose.Console/InventoryFactory.cs]: I had trouble with my `InventoryConditionAdvancerTests` initially when using `yield` syntax in this method.  In my test, I was trying to create the list of default items, make a deep copy of it, run the `AdvanceProperties` method, and then compare the list of changed items to the deep copy to assert they were different.  The way I initially wrote the tests, the list was not being iterated and copied until after the `AdvanceProperties` method was run because of lazy evaluation, so my comparison was finding the list and its copy to be the same.  I created this intermediary list in `CreateDefaultItems`, because it fixed the issue (though I still don't understand completely why).

After some thought, I realized it would be cleaner to force iteration in my tests, which is what I did in commit 9c4ad5d2a5a724038fd3d8651a44b67d39e69c33.  It took me a while to figure this out, and I did some good reading on query expressions during this time.

**Comment by FGS-BMarkham** [comment]
:gopher: :squash:

**Comment by FGS-BMarkham** [comment]
:shipit:

**Comment by FGS-BMarkham** [comment]
Crap. that wasn't pointed at a dedicated branch for Phil. I'll fix `master` for the next recruit…
